### PR TITLE
Make creation of static token kubeconfig for shoot clusters configurable

### LINKED ISSUE #3138: Deactivate the kubeconfig provided in Gardener dashboard
/area security
/kind enhancement
/priority normal

**What would you like to be added**:
Hello gardener team,
please provide the possibility to deactivate the admin-kubeconfig accessible in Gardener dashboard.

**Why is this needed**:
We are already creating & rotating a kubeconfig each 24h with the help of service accounts. Since we do not need the gardener provided kubeconfig, we would like to disable it completely.

### PR DESCRIPTION
**Pull request by ary1992**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
Currently Kubeconfig on gardener dashboard is using static token credentials. New feature gate [`AdminKubeconfigRequest`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md)  was introduced which will enable users to dynamically generate a short-lived kubeconfig that can be used to access the Shoot cluster as cluster-admin. As this feature is moved to `beta` (i.e now it is enabled by default), this PR will make creation of static token kubeconfig configurable .

**Which issue(s) this PR fixes**:
Fixes #3138

**Special notes for your reviewer**:

**Release note**:

```feature user
The static admin kubeconfig for shoot clusters can now be disabled by setting `.spec.kubernetes.enableStaticTokenKubeconfig=false` in the specification of the `Shoot` resource. The respective `<shoot-name>.kubeconfig` secret in the project namespace in the garden cluster will be deleted.
```
```feature user
Project admins can now issue short-lived cluster-admin kubeconfigs by using the `shoots/adminkubeconfig` subresource. Please consult [this document](https://github.com/gardener/gardener/tree/master/example/shoot-kubeconfig) for more information.
```

### CODE REVIEW

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 1]
```diff
@@ -233,6 +235,7 @@ func (k *kubeAPIServer) Deploy(ctx context.Context) error {
 		configMapAdmission                   = k.emptyConfigMap(configMapAdmissionNamePrefix)
 		configMapAuditPolicy                 = k.emptyConfigMap(configMapAuditPolicyNamePrefix)
 		configMapEgressSelector              = k.emptyConfigMap(configMapEgressSelectorNamePrefix)
+		enableAdminKubeconfig                = *(k.GetValues().EnableAdminKubeconfig)
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]:
```suggestion
```

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 2]
```diff
@@ -307,8 +310,10 @@ func (k *kubeAPIServer) Deploy(ctx context.Context) error {
 		return err
 	}
 
-	if err := k.reconcileSecretUserKubeconfig(ctx, secretStaticToken, secretBasicAuth); err != nil {
-		return err
+	if enableAdminKubeconfig {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]:
```suggestion
	if pointer.BoolDeref(k.values.EnableAdminKubeconfig, true) {
```

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 3]
```diff
@@ -307,8 +310,10 @@ func (k *kubeAPIServer) Deploy(ctx context.Context) error {
...
+		if err := k.reconcileSecretUserKubeconfig(ctx, secretStaticToken, secretBasicAuth); err != nil {
+			return err
+		}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go:399]: When `EnableAdminKubeconfig` is `false` then this secret should be deleted. I would implement this in https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/component/kubeapiserver/secrets.go#L131 and not here (than above `if` condition should also be moved to this function).

File: pkg/operation/botanist/kubeapiserver.go [hunk 4]
```diff
@@ -445,28 +446,33 @@ func (b *Botanist) DeployKubeAPIServer(ctx context.Context) error {
 		return err
 	}
 
-	userKubeconfigSecret, found := b.SecretsManager.Get(kubeapiserver.SecretNameUserKubeconfig)
-	if !found {
-		return fmt.Errorf("secret %q not found", kubeapiserver.SecretNameUserKubeconfig)
-	}
+	if *(b.Shoot.GetInfo().Spec.Kubernetes.EnableAdminKubeconfig) {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
	if enableAdminKubeconfig := b.Shoot.GetInfo().Spec.Kubernetes.EnableAdminKubeconfig; enableAdminKubeconfig == nil || *enableAdminKubeconfig {
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 5]
```diff
@@ -445,28 +446,33 @@ func (b *Botanist) DeployKubeAPIServer(ctx context.Context) error {
...
+			gutil.ShootProjectSecretSuffixKubeconfig,
+			map[string]string{v1beta1constants.GardenRole: v1beta1constants.GardenRoleKubeconfig},
+			map[string]string{"url": "https://" + externalServer},
+			kubeconfigSecretData,
+		); err != nil {
+			return err
+		}
+	} else {
+		secretName := gutil.ComputeShootProjectSecretName(b.Shoot.GetInfo().Name, gutil.ShootProjectSecretSuffixKubeconfig)
+		return kutil.DeleteObject(ctx, b.K8sSeedClient.Client(), &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.GetInfo().Namespace}})
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
		if err := kutil.DeleteObject(ctx, b.K8sSeedClient.Client(), &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.GetInfo().Namespace}}); err != nil {
			return err
		}
```

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 6]
```diff
@@ -1529,6 +1543,28 @@ var _ = Describe("KubeAPIServer", func() {
 				HaveKeyWithValue("data-for", []byte("user-kubeconfig")),
 			))
 		})
+
+		It("should not sync the kubeconfig to garden project namespace when enableAdminKubeconfig is set to false", func() {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]: Create the kubeconfig secret in the garden cluster before calling `DeployKubeAPIServer` to improve the test accuracy.

**Review by rfranzke — CHANGES_REQUESTED** [review]
This https://github.com/gardener/gardener/blob/3d0d818e213eec89d225d71aa6bf574b48c4b7cc/pkg/operation/botanist/component/kubeapiserver/secrets.go#L112-L116
should not even be there when the admin kubeconfig is disabled.

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go (hunk 3 above)
Comment (ary1992) [reply pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go:399]: I have tested it on local. The secret is getting deleted when EnableAdminKubeconfig is set to false. I thought it will good we don't call the `reconcileSecretUserKubeconfig` function entirely when `enableAdminKubeconfig ` is set to false. WDYT ?

Comment (rfranzke) [reply pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go:399]: Ah, nice, this is happening because we call `Cleanup` of secrets manager. Yeah, all good then, nothing to be done. Keep it as is!

**Comment by ary1992** [comment]
> This
> 
> https://github.com/gardener/gardener/blob/3d0d818e213eec89d225d71aa6bf574b48c4b7cc/pkg/operation/botanist/component/kubeapiserver/secrets.go#L112-L116
> 
> 
> should not even be there when the admin kubeconfig is disabled.

Similar to `reconcileSecretUserKubeconfig` function , can we also call `reconcileSecretStaticToken` only when `EnableAdminKubeconfig` is set to true i.e move [kube_apiserver.go#L277](https://github.com/gardener/gardener/blob/371a2c95bf17afab16754dd3cc4616fa2262b2b4/pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go#L277) to [here](https://github.com/gardener/gardener/blob/371a2c95bf17afab16754dd3cc4616fa2262b2b4/pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go#L314) ?

**Comment by rfranzke** [comment]
@ary1992 No, we need the static token since it also includes the token for the health check.

File: pkg/operation/botanist/kubeapiserver.go [hunk 7]
```diff
@@ -445,28 +446,35 @@ func (b *Botanist) DeployKubeAPIServer(ctx context.Context) error {
...
+			map[string]string{v1beta1constants.GardenRole: v1beta1constants.GardenRoleKubeconfig},
+			map[string]string{"url": "https://" + externalServer},
+			kubeconfigSecretData,
+		); err != nil {
+			return err
+		}
+	} else {
+		secretName := gutil.ComputeShootProjectSecretName(b.Shoot.GetInfo().Name, gutil.ShootProjectSecretSuffixKubeconfig)
+		if err := kutil.DeleteObject(ctx, b.K8sSeedClient.Client(), &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.GetInfo().Namespace}}); err != nil {
+			return nil
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
			return err
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 8]
```diff
@@ -445,28 +446,35 @@ func (b *Botanist) DeployKubeAPIServer(ctx context.Context) error {
...
+			gutil.ShootProjectSecretSuffixKubeconfig,
+			map[string]string{v1beta1constants.GardenRole: v1beta1constants.GardenRoleKubeconfig},
+			map[string]string{"url": "https://" + externalServer},
+			kubeconfigSecretData,
+		); err != nil {
+			return err
+		}
+	} else {
+		secretName := gutil.ComputeShootProjectSecretName(b.Shoot.GetInfo().Name, gutil.ShootProjectSecretSuffixKubeconfig)
+		if err := kutil.DeleteObject(ctx, b.K8sSeedClient.Client(), &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.GetInfo().Namespace}}); err != nil {
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]:
```suggestion
		if err := kutil.DeleteObject(ctx, b.K8sGardenClient.Client(), &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.GetInfo().Namespace}}); err != nil {
```

File: pkg/operation/botanist/kubeapiserver_test.go (hunk 6 above)
Comment (rfranzke) [reply pkg/operation/botanist/kubeapiserver_test.go]: This is not yet resolved

File: pkg/operation/botanist/component/kubeapiserver/secrets.go [hunk 9]
```diff
@@ -104,20 +105,25 @@ func (k *kubeAPIServer) reconcileSecretBasicAuth(ctx context.Context) (*corev1.S
...
 			userNameHealthCheck: {
 				Username: userNameHealthCheck,
 				UserID:   userNameHealthCheck,
 			},
 		},
-	}, secretsmanager.Persist(), secretsmanager.Rotate(secretsmanager.InPlace))
+	}
+
+	if pointer.BoolDeref(k.values.EnableAdminKubeconfig, true) {
+		staticTokenSecretConfig.Tokens["userNameClusterAdmin"] = secretutils.TokenConfig{
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/secrets.go]:
```suggestion
		staticTokenSecretConfig.Tokens[userNameClusterAdmin] = secretutils.TokenConfig{
```

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go [hunk 10]
```diff
@@ -307,8 +309,10 @@ func (k *kubeAPIServer) Deploy(ctx context.Context) error {
 		return err
 	}
 
-	if err := k.reconcileSecretUserKubeconfig(ctx, secretStaticToken, secretBasicAuth); err != nil {
-		return err
+	if pointer.BoolDeref(k.values.EnableAdminKubeconfig, true) {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]: I am missing a unit test case for this behaviour.

File: pkg/operation/botanist/component/kubeapiserver/secrets.go [hunk 11]
```diff
@@ -104,20 +105,25 @@ func (k *kubeAPIServer) reconcileSecretBasicAuth(ctx context.Context) (*corev1.S
...
-			},
 			userNameHealthCheck: {
 				Username: userNameHealthCheck,
 				UserID:   userNameHealthCheck,
 			},
 		},
-	}, secretsmanager.Persist(), secretsmanager.Rotate(secretsmanager.InPlace))
+	}
+
+	if pointer.BoolDeref(k.values.EnableAdminKubeconfig, true) {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/secrets.go]: I am missing a unit test case for this behaviour.

**Review by rfranzke — CHANGES_REQUESTED** [review]
Apart from the feedback above, please also extend the project-member `ClusterRole` (https://github.com/gardener/gardener/blob/55de9acff4d85d4fcf6ba716b4273186abd92788/charts/gardener/controlplane/charts/application/templates/rbac-user.yaml#L289-L387) and allow to `CREATE` `shoots/adminkubeconfig` so that project admins can issue themselves a temporary kubeconfig when they have deactivated the static admin kubeconfig.

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go (hunk 10 above)
Comment (ary1992) [reply pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]: Is [this](https://github.com/gardener/gardener/blob/18fa04ac2710ab344dddf6378ad36c87d9a5dea1/pkg/operation/botanist/component/kubeapiserver/kube_apiserver_test.go#L1958) not the correct unit test ?

Comment (rfranzke) [reply pkg/operation/botanist/component/kubeapiserver/kube_apiserver.go]: Right, sorry, I missed this case

**Comment by rfranzke** [comment]
cc @grolu @holgerkoser @petersutter for needed dashboard adaptations

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver_test.go [hunk 12]
```diff
@@ -1998,8 +2002,20 @@ rules:
 						Expect(kubeconfig.AuthInfos[0].AuthInfo.Token).NotTo(BeEmpty())
 					})
 
+					It("should not generate a kubeconfig secret for the user when EnableAdminKubeconfig is set to false", func() {
```
Comment (acumino) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver_test.go:2005]: Since EnableAdminKubeconfig is optional can we add one more test where if it not passed it should be defaulted to true?

File: pkg/operation/botanist/component/kubeapiserver/kube_apiserver_test.go [hunk 13]
```diff
@@ -2242,7 +2315,7 @@ rules:
 					})
 
 					It("should not configure the settings related to the basic auth secret if disabled", func() {
-						kapi = New(kubernetesInterface, namespace, sm, Values{Images: images, Version: version, BasicAuthenticationEnabled: false})
+						kapi = New(kubernetesInterface, namespace, sm, Values{Images: images, Version: version, BasicAuthenticationEnabled: false, EnableAdminKubeconfig: enableAdminKubeconfig})
```
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserver/kube_apiserver_test.go]: Why do you have to set this `EnableAdminKubeconfig` value for all the occurrences of `New` here? Can you remove this again? The code should also work when this is `nil`, right?

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 14]
```diff
@@ -1269,6 +1274,7 @@ var _ = Describe("KubeAPIServer", func() {
 									MaxTokenExpiration:    &maxTokenExpiration,
 								},
 							},
+							EnableAdminKubeconfig: enableAdminKubeconfig,
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]: Same here, why is this needed? Remove it please :)

**Review by rfranzke — COMMENTED** [review]
Wow, very nice PR, I'm very happy with it! Thanks a lot. Just two minor suggestions :)

**Review by rfranzke — COMMENTED** [review]
/lgtm
/milestone v1.44
/approve

**Review by rfranzke — APPROVED** [review]
